File: backend/app/realtime/connection_manager.py
from __future__ import annotations

from collections.abc import Iterable
from collections import defaultdict
from dataclasses import dataclass

from fastapi import WebSocket, WebSocketDisconnect


@dataclass(slots=True)
class RoomConnection:
    target_lang: str = "en"


class ConnectionManager:
    def __init__(self) -> None:
        self._rooms: dict[str, dict[WebSocket, RoomConnection]] = defaultdict(dict)

    async def connect(self, conversation_id: str, websocket: WebSocket, *, target_lang: str = "en") -> None:
        await websocket.accept()
        self._rooms[conversation_id][websocket] = RoomConnection(target_lang=target_lang)

    def disconnect(self, conversation_id: str, websocket: WebSocket) -> None:
        room = self._rooms.get(conversation_id)
        if not room:
            return
        room.pop(websocket, None)
        if not room:
            self._rooms.pop(conversation_id, None)

    def update_target_lang(self, conversation_id: str, websocket: WebSocket, target_lang: str) -> None:
        room = self._rooms.get(conversation_id)
        if not room or not target_lang:
            return
        connection = room.get(websocket)
        if connection is None:
            return
        connection.target_lang = target_lang

    def snapshot_target_groups(self, conversation_id: str) -> dict[str, tuple[WebSocket, ...]]:
        grouped: dict[str, list[WebSocket]] = defaultdict(list)
        for websocket, connection in self._rooms.get(conversation_id, {}).items():
            grouped[connection.target_lang].append(websocket)
        return {target_lang: tuple(websockets) for target_lang, websockets in grouped.items()}

    async def broadcast(self, conversation_id: str, payload: dict) -> None:
        await self.send(conversation_id, tuple(self._rooms.get(conversation_id, {})), payload)

    async def send(self, conversation_id: str, websockets: Iterable[WebSocket], payload: dict) -> None:
        for websocket in tuple(websockets):
            try:
                await websocket.send_json(payload)
            except (RuntimeError, WebSocketDisconnect):
                self.disconnect(conversation_id, websocket)

File: go-app/internal/realtime/connection_manager.go
package realtime

import (
	"sync"

	"github.com/gorilla/websocket"
)

type RoomConnection struct {
	TargetLang string
}

type ConnectionManager struct {
	mu    sync.Mutex
	rooms map[string]map[*websocket.Conn]RoomConnection
}

func NewConnectionManager() *ConnectionManager {
	return &ConnectionManager{rooms: map[string]map[*websocket.Conn]RoomConnection{}}
}

func (m *ConnectionManager) Connect(conversationID string, conn *websocket.Conn, targetLang string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if targetLang == "" {
		targetLang = "en"
	}
	room := m.rooms[conversationID]
	if room == nil {
		room = map[*websocket.Conn]RoomConnection{}
		m.rooms[conversationID] = room
	}
	room[conn] = RoomConnection{TargetLang: targetLang}
}

func (m *ConnectionManager) Disconnect(conversationID string, conn *websocket.Conn) {
	m.mu.Lock()
	defer m.mu.Unlock()
	room := m.rooms[conversationID]
	if room == nil {
		return
	}
	delete(room, conn)
	if len(room) == 0 {
		delete(m.rooms, conversationID)
	}
}

func (m *ConnectionManager) UpdateTargetLang(conversationID string, conn *websocket.Conn, targetLang string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	room := m.rooms[conversationID]
	if room == nil || targetLang == "" {
		return
	}
	state, ok := room[conn]
	if !ok {
		return
	}
	state.TargetLang = targetLang
	room[conn] = state
}

func (m *ConnectionManager) SnapshotTargetGroups(conversationID string) map[string][]*websocket.Conn {
	m.mu.Lock()
	defer m.mu.Unlock()
	grouped := map[string][]*websocket.Conn{}
	for conn, state := range m.rooms[conversationID] {
		grouped[state.TargetLang] = append(grouped[state.TargetLang], conn)
	}
	return grouped
}

func (m *ConnectionManager) Send(conversationID string, conns []*websocket.Conn, payload map[string]any) {
	for _, conn := range conns {
		if err := conn.WriteJSON(payload); err != nil {
			m.Disconnect(conversationID, conn)
			conn.Close()
		}
	}
}
